### PR TITLE
Expose the Entity Counter to allow plugins to use valid and non-conflicting Entity Ids

### LINKED ISSUE #4251: Entity Id generator API
Right now any plugin that wants to implement network only entities has to reflect against entity id generator field. While network only plugins usually require updating to new nms version each release anyway, having some sort of API to at least being able to avoid at least reflection this field would be great.

### PR DESCRIPTION
**Pull request by MeFisto94**
Basically without this (or reflection), people would have to send random (or very high value entity ids) for their packets, because otherwise "ghost events" get fired on the server for the completely wrong entities and other strange things might happen to the client (teleporting entities).

Using EntityIds provides the only safe way to have no collisions with regular server entities and also makes it easy for the plugin to track which packet-only entities it had.

### CODE REVIEW

**Review by MiniDigger — DISMISSED** [review]
looks fine to me

File: Spigot-API-Patches/0222-Expose-the-Entity-Counter-to-allow-plugins-to-use-va.patch
Comment (kashike) [inline Spigot-API-Patches/0222-Expose-the-Entity-Counter-to-allow-plugins-to-use-va.patch]: Perhaps `allocateEntityId`?

**Review by kashike — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0222-Expose-the-Entity-Counter-to-allow-plugins-to-use-va.patch
Comment (MeFisto94) [reply Spigot-API-Patches/0222-Expose-the-Entity-Counter-to-allow-plugins-to-use-va.patch]: The problem might be that it has the feeling of being able to "free" the entityId again (but `create` is no exception there).
But I guess any formulation will suffer from that

Comment (yannicklamprecht) [reply Spigot-API-Patches/0222-Expose-the-Entity-Counter-to-allow-plugins-to-use-va.patch]: What is with nextEntityId() ?

Comment (MeFisto94) [reply Spigot-API-Patches/0222-Expose-the-Entity-Counter-to-allow-plugins-to-use-va.patch]: nextEntityId sounds good to me, especially as it indicates that it's somehow fetching id and doesn't have the `alloc/free` notion :)

Comment (kashike) [reply Spigot-API-Patches/0222-Expose-the-Entity-Counter-to-allow-plugins-to-use-va.patch]: `nextEntityId` works for me.

**Comment by MeFisto94** [comment]
Looks like I messed something up, as in the initial patch da33f8bbe106309055d340c868c1138801536b88 was not present.
I think I'd need to rebase the patches to master after applying and then rebuilding them?

**Comment by MiniDigger** [comment]
you basically checkout master and pull, to make sure its up to date, checkout your branch `git rebase master`, reorder your patch to be the last, `paper p && paper rb`.

you might need to fix conflicts when patching. 

but you can also leave it as it and we will handle it for you, once there is an approval by a core team member.

and well, it looks like you figured it out, so nvm.

File: Spigot-API-Patches/0222-Expose-the-Entity-Counter-to-allow-plugins-to-use-va.patch
Comment (LuminousLoreBloom) [reply Spigot-API-Patches/0222-Expose-the-Entity-Counter-to-allow-plugins-to-use-va.patch]: I just thought of getAndIncrementEntityId to match AtomicInteger but this is such a trivial thing that's it not worth editing for

**Comment by MeFisto94** [comment]
Does this require anything? i.e. another rebase?

**Review by kashike — DISMISSED** [review]
Looks good to me.

**Comment by Fejm** [comment]
Well.. You can just create your own counter starting with Integer.MIN_VALUE and leave internal entity counter alone..

**Comment by kennytv** [comment]
now imagine 2 plugins doing the same. using the one, proper counter is the only thing you should do for custom entity magic.

**Comment by Draycia** [comment]
The point is to prevent conflicts, plugins using their own counter is precisely the thing we're trying to avoid.  
That method is how conflicts occur.  
As per the OP, `because otherwise "ghost events" get fired on the server for the completely wrong entities`.  

Essentially those two plugins would be controlling / modifying each other's entities, as Entity IDs are global and are NOT per plugin.

**Comment by Fejm** [comment]
Then create API in Paper like suggested but using custom counter. But this is only my opinion.

**Comment by Draycia** [comment]
Again, Entity IDs are global for the server.  
Plugin specific counters address nothing, because you could still conflict with vanilla.

**Comment by Fejm** [comment]
Vanilla uses positive values for Enitity ID...

Use negative for custom needs. I understood what you say. But if you are afraid of ID collision then PaperAPI would provide separate counter starting at Integer.MIN_VALUE to disting custom ID from vanilla ID.

**Comment by Shevchik** [comment]
Vanilla uses always incrementing counter, which will overflow to negative values.

**Review by aikar — DISMISSED** [review]
I don't know why you're so opposed to this but this is the only valid solution to this problem, so this is fine as is.

**Review by electronicboy — APPROVED** [review]

**Comment by MiniDigger** [comment]
I don't trust @electronicboy's review!!!!
limbo, do your thing!
@limbo-app ready

**Review by limbo-app — APPROVED** [review]
Pull request automatically approved for merging - a core team member has previously approved this pull request.

**Comment by MiniDigger** [comment]
thanks for the PR!

**Pull request merged**
